File: src/main/java/net/snipsniper/utils/FileUtils.java
package net.snipsniper.utils;

import org.capturecoop.cclogger.CCLogger;
import org.capturecoop.cclogger.CCLogLevel;
import org.capturecoop.ccutils.utils.CCStringUtils;

import java.awt.*;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;

public class FileUtils {

    public static boolean deleteRecursively(String folder) {
        return deleteRecursively(new File(folder));
    }

    public static boolean deleteRecursively(File folder) {
        if(!folder.isDirectory()) {
            return FileUtils.delete(folder);
        }

        boolean success = true;
        for(File file : FileUtils.listFiles(folder)) {
            if(file.isDirectory()) {
                if(!deleteRecursively(file))
                    success = false;
            }
            if(!FileUtils.delete(file))
                success = false;
        }
        if(!FileUtils.delete(folder))
            success = false;
        return success;
    }

    public static boolean delete(String file) {
        return delete(new File(file));
    }

    public static boolean delete(File file) {
        if(!file.exists()) return true;

        if(!file.delete()) {
            CCLogger.Companion.warn("File (" + file.getAbsolutePath() + ") could not be deleted!");
            return false;
        }
        return true;
    }

    public static boolean mkdir(String file) {
        return mkdir(new File(file));
    }

    public static boolean mkdir(File folder) {
        if(folder.exists()) return true;

        if(!folder.mkdir()) {
            CCLogger.Companion.warn("Folder (" + folder.getAbsolutePath() + ") could not be created.");
            return false;
        }
        return true;
    }

    public static boolean mkdirs(String file) {
        return mkdirs(new File(file));
    }

    public static boolean mkdirs(File folder) {
        if(folder.exists()) return true;

        if(!folder.mkdirs()) {
            CCLogger.Companion.warn("Folders (" + folder + ") could not be created.");
            return false;
        }
        return true;
    }

    public static boolean mkdirs(String... folders) {
        File[] array = new File[folders.length];
        for(int i = 0; i < folders.length; i++)
            array[i] = new File(folders[i]);
        return mkdirs(array);
    }

    public static boolean mkdirs(File... folders) {
        boolean success = true;
        for(File folder : folders)
            if(!mkdirs(folder))
                success = false;
        return success;
    }

    public static File[] listFiles(String folder) {
        return listFiles(new File(folder));
    }

    public static File[] listFiles(File folder) {
        return folder.listFiles();
    }

    public static ArrayList<String> getFilesInFolders(String path) {
        ArrayList<String> result = new ArrayList<>();
        for(File file : FileUtils.listFiles(path)) {
            if(file.isDirectory())
                result.addAll(getFilesInFolders(file.getAbsolutePath()));
            if(!file.isDirectory())
                result.add(CCStringUtils.correctSlashes(file.getAbsolutePath()));
        }
        return result;
    }

    public static boolean exists(String... files) {
        File[] f = new File[files.length];
        for(int i = 0; i < files.length; i++)
            f[i] = new File(files[i]);
        return exists(f);
    }

    public static boolean exists(File... files) {
        boolean allExist = true;
        for(File file : files) {
            if(!file.exists())
                allExist = false;
        }
        return allExist;
    }

    public static void openFolder(String path) {
        try {
            Desktop.getDesktop().open(new File(path));
        } catch (IOException ioException) {
            CCLogger.Companion.error("Could not open folder \"" + path + "\"!");
            CCLogger.Companion.logStacktrace(ioException, CCLogLevel.ERROR);
        }
    }

    public static void printFile(String filename, String text) {
        try {
            PrintWriter out = new PrintWriter(filename);
            out.print(text);
            out.close();
        } catch (FileNotFoundException fileNotFoundException) {
            CCLogger.Companion.error("Could not write to file \"" + filename + "\"!");
            CCLogger.Companion.logStacktrace(fileNotFoundException, CCLogLevel.ERROR);
        }
    }

    public static String getFileExtension(File file) {
        return getFileExtension(file, true);
    }

    public static String getFileExtension(File file, boolean dot) {
        String name = file.getName();
        int lastIndexOf = name.lastIndexOf(".");
        if (lastIndexOf == -1) {
            return ""; // empty extension
        }
        if(!dot)
            lastIndexOf++;
        return name.substring(lastIndexOf);
    }

    public static String getCanonicalPath(String path) {
        try {
            return new File(path).getCanonicalPath();
        } catch (IOException ioException) {
            CCLogger.Companion.error("Could not get path for \"" + path + "\"!");
            CCLogger.Companion.logStacktrace(ioException, CCLogLevel.ERROR);
        }
        return null;
    }

    public static boolean copyFromJar(String jarPath, String path) {
        if(jarPath.startsWith("\\") || jarPath.startsWith("//"))
            CCLogger.Companion.warn("jarPath is starting with slashes, this generally does not work inside the jar!");

        if(FileUtils.exists(path))
            FileUtils.delete(path);

        InputStream inputStream = ClassLoader.getSystemResourceAsStream(jarPath);
        if(inputStream == null) {
            CCLogger.Companion.error(String.format("InputStream is null! Copying failed! jarPath: %s, path: %s", jarPath, path));
            return false;
        }
        try {
            Files.copy(inputStream, new File(path).getCanonicalFile().toPath(), StandardCopyOption.REPLACE_EXISTING);
        } catch (IOException ioException) {
            CCLogger.Companion.log("Issue copying from jar!", CCLogLevel.ERROR);
            CCLogger.Companion.logStacktrace(ioException, CCLogLevel.ERROR);
            return false;
        }
        return true;
    }

    public static String loadFileFromJar(String file) {
        StringBuilder content = new StringBuilder();
        try{
            String path = "net/snipsniper/resources/" + file;
            InputStream inputStream = ClassLoader.getSystemResourceAsStream(path);
            if(inputStream == null) {
                CCLogger.Companion.log(CCStringUtils.format("Could not load file %c from jar!", path), CCLogLevel.ERROR);
                return null;
            }
            InputStreamReader streamReader = new InputStreamReader(inputStream, StandardCharsets.UTF_8);
            BufferedReader in = new BufferedReader(streamReader);

            for (String line; (line = in.readLine()) != null;)
                content.append(line);

            inputStream.close();
            streamReader.close();
        } catch (IOException ioException) {
            CCLogger.Companion.log("Could not load file: " + file, CCLogLevel.ERROR);
            CCLogger.Companion.logStacktrace(ioException, CCLogLevel.ERROR);
        }
        return content.toString();
    }
}

File: src/main/kotlin/net/snipsniper/utils/FileUtils.kt
package net.snipsniper.utils

import net.snipsniper.SnipSniper
import org.capturecoop.defaultdepot.StringUtils
import org.capturecoop.legiblelogger.LegibleLogLevel
import org.capturecoop.legiblelogger.LegibleLogger
import java.awt.Desktop
import java.io.*
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Paths
import java.nio.file.StandardCopyOption

class FileUtils {
    companion object {
        fun delete(file: String) = delete(File(file))

        fun delete(file: File): Boolean {
            if (!file.exists()) return true
            File("").delete()
            if (!file.delete()) {
                LegibleLogger.warn("File (${file.absolutePath}) could not be deleted!")
                return false
            }
            return true
        }

        fun mkdir(file: String) = mkdir(File(file))

        fun mkdir(folder: File): Boolean {
            if (folder.exists()) return true
            if (!folder.mkdir()) {
                LegibleLogger.warn("Folder (${folder.absolutePath}) could not be created.")
                return false
            }
            return true
        }

        fun mkdirs(file: String?) = mkdirs(File(file))

        fun mkdirs(folder: File): Boolean {
            if (folder.exists()) return true
            if (!folder.mkdirs()) {
                LegibleLogger.warn("Folders ($folder) could not be created.")
                return false
            }
            return true
        }

        fun mkdirs(vararg folders: String): Boolean {
            return mkdirs(*Array(folders.size) { i -> File(folders[i]) })
        }

        fun mkdirs(vararg folders: File): Boolean {
            var success = true
            for (folder in folders) if (!mkdirs(folder)) success = false
            return success
        }

        fun listFiles(folder: String) = listFiles(File(folder))

        fun listFiles(folder: File) = folder.listFiles()

        fun getFilesInFolders(path: String): ArrayList<String> {
            val result = ArrayList<String>()
            for (file in listFiles(path)) {
                if (file.isDirectory) result.addAll(getFilesInFolders(file.absolutePath))
                if (!file.isDirectory) result.add(StringUtils.correctSlashes(file.absolutePath))
            }
            return result
        }

        fun exists(vararg files: String) = exists(*Array(files.size) { i -> File(files[i]) })

        fun exists(vararg files: File): Boolean {
            var allExist = true
            for (file in files) {
                if (!file.exists()) allExist = false
            }
            return allExist
        }

        fun openFolder(path: String) {
            try {
                Desktop.getDesktop().open(File(path))
            } catch (ioException: IOException) {
                LegibleLogger.error("Could not open folder \"$path\"!")
                LegibleLogger.logStacktrace(ioException, LegibleLogLevel.ERROR)
            }
        }

        fun printFile(filename: String, text: String?) {
            try {
                PrintWriter(filename).also {
                    it.print(text)
                    it.close()
                }
            } catch (fileNotFoundException: FileNotFoundException) {
                LegibleLogger.error("Could not write to file \"$filename\"!")
                LegibleLogger.logStacktrace(fileNotFoundException, LegibleLogLevel.ERROR)
            }
        }

        fun getFileExtension(file: File): String {
            return getFileExtension(file, true)
        }

        fun getFileExtension(file: File, dot: Boolean): String {
            val name = file.name
            var lastIndexOf = name.lastIndexOf(".")
            if (lastIndexOf == -1) {
                return "" // empty extension
            }
            if (!dot) lastIndexOf++
            return name.substring(lastIndexOf)
        }

        fun getCanonicalPath(path: String): String? {
            try {
                return File(path).canonicalPath
            } catch (ioException: IOException) {
                LegibleLogger.error("Could not get path for \"$path\"!")
                LegibleLogger.logStacktrace(ioException, LegibleLogLevel.ERROR)
            }
            return null
        }

        fun getJarFolder(): String {
            URLDecoder.decode(Paths.get(SnipSniper::class.java.protectionDomain.codeSource.location.toURI()).toString(), "UTF-8")!!.also { fullPath ->
                File(fullPath).also { file ->
                    return if(file.isFile) fullPath.replace(file.name, "") else fullPath
                }
            }
        }

        fun copyFromJar(jarPath: String, path: String): Boolean {
            if (jarPath.startsWith("\\") || jarPath.startsWith("//"))
                LegibleLogger.warn("jarPath ($path) is starting with slashes, this generally does not work inside the jar!")
            if (exists(path))
                delete(path)
            val inputStream = ClassLoader.getSystemResourceAsStream(jarPath)
            if (inputStream == null) {
                LegibleLogger.error("InputStream is null! Copying failed! jarPath: $jarPath, path: $path")
                return false
            }
            try {
                Files.copy(inputStream, File(path).canonicalFile.toPath(), StandardCopyOption.REPLACE_EXISTING)
            } catch (ioException: IOException) {
                LegibleLogger.error("Issue copying from jar!")
                LegibleLogger.logStacktrace(ioException, LegibleLogLevel.ERROR)
                return false
            }
            return true
        }

        fun loadFileFromJar(file: String): String? {
            val content = StringBuilder()
            try {
                val path = "net/snipsniper/resources/$file"
                val inputStream = ClassLoader.getSystemResourceAsStream(path)
                if(inputStream == null) {
                    LegibleLogger.error("Could not load file $path from jar!")
                    return null
                }
                val streamReader = InputStreamReader(inputStream, StandardCharsets.UTF_8)
                BufferedReader(streamReader).also {
                    it.readLines().forEach { line -> content.append(line) }
                    it.close()
                }
                streamReader.close()
            } catch (ioException: IOException) {
                LegibleLogger.error("Could not load file: $file")
                LegibleLogger.logStacktrace(ioException, LegibleLogLevel.ERROR)
            }
            return content.toString()
        }
    }
}